api/src/main/java/org/openmrs/api/impl/CohortServiceImpl.java

```diff
@@ -57,108 +58,141 @@ public Cohort saveCohort(Cohort cohort) throws APIException {
-			cohort.getMemberIds().remove(patient.getPatientId());
+		if (cohort.contains(patient.getPatientId())) {
+            cohort.removeMember(patient, "Removing patient from Cohort");
 			Context.getCohortService().saveCohort(cohort);
 		}
 		return cohort;
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohortsContainingPatient(org.openmrs.Patient)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohortsContainingPatient(Patient patient) {
 		return dao.getCohortsContainingPatientId(patient.getPatientId());
 	}
-	
+
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohortsContainingPatientId(Integer patientId) {
 		return dao.getCohortsContainingPatientId(patientId);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohorts(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohorts(String nameFragment) throws APIException {
 		return dao.getCohorts(nameFragment);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts()
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts() throws APIException {
 		return Context.getCohortService().getAllCohorts(false);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts(boolean)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts(boolean includeVoided) throws APIException {
 		return dao.getAllCohorts(includeVoided);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohort(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public Cohort getCohort(String name) throws APIException {
 		return dao.getCohort(name);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#purgeCohort(org.openmrs.Cohort)
 	 */
 	public Cohort purgeCohort(Cohort cohort) throws APIException {
 		return dao.deleteCohort(cohort);
 	}
-	
+
+	/**
+	 * @see org.openmrs.api.CohortService#addMembershipToCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort addMembershipToCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+        cohort.addMembership(cohortMembership);
+		Context.getCohortService().saveCohort(cohort);
+		return cohort;
+	}
+
+	/**
+	 * @see org.openmrs.api.CohortService#removeMemberShipFromCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort removeMemberShipFromCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+		if (cohort.contains(cohortMembership)) {
+			cohort.removeMembership(cohortMembership);
+			Context.getCohortService().saveCohort(cohort);
+		}
+		return cohort;
+	}
+
+	/**
+     * @see org.openmrs.api.CohortService#patientVoided(org.openmrs.Patient)
+	 */
+	public void patientVoided(Patient patient) throws APIException {
+        List<Cohort> cohorts = Context.getCohortService().getCohortsContainingPatient(patient);
+        for (Cohort cohort : cohorts) {
+			cohort.removeMember(patient, "Patient is voided");
```

Comment:
I did come across this situation. So if the patient has it's dateVoided, voidedBy etc. set. Could I set the membership void details the same as the patient in removeMember?